File: HarpiaTest/app_example/fanout/java/src/main/java/com/harpia/fanout/Worker.java
// Java variant of the PUSH/PULL load-balance worker (transport-multipeer-
// coverage task 6) -- proves a Java copy takes a real share of the work
// alongside any number of the C++ `worker` binary, the same cross-language
// interoperability UnitTests/test_zmq_xlang_pushpull_loadbalance.py proves
// structurally. See ../../src/worker.cpp for the C++ twin and
// ../../README.md for how to run this alongside it.
//
//   java -cp <classpath> com.harpia.fanout.Worker [endpoint]
//     endpoint  default tcp://127.0.0.1:5571 -- this copy's OWN port, must
//               also be listed on `pusher`'s argv
package com.harpia.fanout;

import com.harpia.generated.courier;
import com.harpia.generated.zmq.courier_zmq;
import com.harpia.runtime.zmq.HarpiaZmq;
import org.zeromq.ZContext;

public class Worker {
    public static void main(String[] args) throws Exception {
        String endpoint = args.length > 0 ? args[0] : "tcp://127.0.0.1:5571";

        try (ZContext ctx = new ZContext()) {
            HarpiaZmq.Receiver rcv = courier_zmq.newReceiver(ctx, endpoint);
            rcv.socket().setReceiveTimeOut(10000);
            System.out.println("worker: bound to " + endpoint);

            for (;;) {
                courier.Builder b = courier.newBuilder();
                if (!rcv.receive(b)) {
                    System.out.println("worker: no work for 10s, exiting");
                    break;
                }
                System.out.println("worker: processed " + b.getPayload());
            }
        }
    }
}

File: HarpiaTest/app_example/fanout/src/worker.cpp
// Worker half of the PUSH/PULL load-balance worked example (transport-
// multipeer-coverage task 6). Each copy needs its OWN endpoint -- a bound
// PULL socket can't share a port with another copy -- so pass it a
// distinct port and list every worker's endpoint on `pusher`'s argv.
// Exits after 10s of silence, so it also works as a bounded, scriptable
// check rather than only an interactive demo.
//
//   ./worker [endpoint]
//     endpoint  default tcp://127.0.0.1:5571
#include "zmq/courier_3ac5d8b36fc7dcfb70888145147ddfb7_zmq.h"
#include <cstdio>
#include <string>

int main(int argc, char** argv) {
    std::string endpoint = argc > 1 ? argv[1] : "tcp://127.0.0.1:5571";

    ::zmq::context_t ctx{1};
    harpia::zmq_transport::courier_receiver rcv(ctx, endpoint);
    rcv.socket().set(::zmq::sockopt::rcvtimeo, 10000);
    std::printf("worker: bound to %s\n", endpoint.c_str());
    std::fflush(stdout);

    for (;;) {
        ::courier in;
        if (!rcv.recv(&in)) {
            std::printf("worker: no work for 10s, exiting\n");
            break;
        }
        std::printf("worker: processed %s\n", in.payload().c_str());
        std::fflush(stdout);
    }
    return 0;
}
